### DISCUSSION TITLE
Max Number of UEs per AGW (with beefy HW)

### DISCUSSION THREAD

**Discussion by jpad-freedomfi** [magma/magma#5507] [Q&A] [answered]
Hi All, 

We had a request to deploy AGW on a beefy server, have multiple ends plumbed to a single AGW instance. The customer was asking what is the maximum number of subscribers a single AGW instance can handle? 

Here is proposed server spec:
```
PowerEdge R340 Server
Intel Xeon E-2234 3.6GHz, 8M cache, 4C/8T, turbo (71W)
PCIe Riser, 1x FH x8 PCIe Gen3 slot, 1x LP x4 PCIe Gen3 slot, R240/R340
Performance Optimized
2666MT/s UDIMMs
16GB 2666MT/s DDR4 ECC UDIMM
iDrac9, Express
480GB SSD SATA Mix Use 6Gbps 512 2.5in Hot-plug AG Drive, 3 DWPD, 2628 TBW
PERC H730P Adapter RAID Controller, 2GB
Heatsink for 80W or less CPU
Dual Hot Plug Power Supplies 350W
Broadcom 57412 Dual Port 10GbE SFP+ Adapter, PCIe Full Height
On-Board LOM
No RAID
```

In slack we heard AGW code has been tested with up to 4k subscribers on a single instance (again with beefy HW).

Does anyone know what the limiting factor becomes? Is it NAT table issues or does something else eventually limit subscriber count? 

Thanks,
-Joey

**Comment by amarpad** [comment, marked as answer]
we have tested 5k for CWF I believe. cc @sudhikan

**Reply by sudhikan** [reply]
Currently we haven't exceeded 600 subs on the AGW as we are trying to nail down a # for a non-beefy hardware (vm or bm). Our target is to double this on the 4G 4-core. 5K was tested in the virtualized CWAG env. and there we were CPU restricted for sure.

**Comment by pshelar** [comment]
On datapath side, current limiting factor is going to be single CPU core performance. datapath does  not always process packets in parallel. We have plans to improve it for R1.6.

**Reply by edaspb** [reply]
Hi @pshelar, I'm curios why it happens? Do you have any details?

**Reply by pshelar** [reply]
First issue is lack of RSS for RTP packet. we need to experiment with different UDP ports.
Other issues are related lock scalability in networking stack, related to TC/QoS.

**Reply by edaspb** [reply]
Thanks. How do you think, if network card queues statically redistribute  between CPU cores will it improve datapath performance?
Attach network card int﻿errupts﻿﻿ to physical cores, something like that:
/proc/int﻿errupts﻿﻿﻿:
```
 26:          3          0          0          0          0          0   PCI-MSI-edge      eth1-rx-0
 27:          0          3          0          0          0          0   PCI-MSI-edge      eth1-rx-1
 28:          0          0          3          0          0          0   PCI-MSI-edge      eth1-rx-2
 29:          0          0          0          3          0          0   PCI-MSI-edge      eth1-tx-0
 30:          0          0          0          0          3          0   PCI-MSI-edge      eth1-tx-1
 31:          0          0          0          0          0          3   PCI-MSI-edge      eth1-tx-2
```
